docs/concepts/flink-architecture.md

```diff
@@ -129,4 +167,108 @@ two main benefits:
+  submits the job to the Dispatcher running inside this process. TaskManagers
+  are then lazily allocated based on the resource requirements of the job. Once
+  the job is finished, the Flink Job Cluster is torn down.
+
+* **Resource Isolation**: a fatal error in the Flink Master only ever affects
+  one job in a Flink Job Cluster.
+
+* **Other considerations**: because the ResourceManager has to apply and wait
+  for external resource management components to start the TaskManager
+  processes and allocate resources, Flink Job Clusters are more suited to large
+  jobs that are long-running, have high-stability requirements and are not
+  sensitive to higher startup times.
+
+<div class="alert alert-info"> <strong>Note:</strong> Formerly, a Flink Job Cluster was also known as a Flink Cluster in <i>job (or per-job) mode</i>. </div>
+
+### Flink Application Cluster
+
+* **Cluster Lifecycle**: a Flink Application Cluster is a dedicated Flink
+  cluster that only executes jobs from one Flink Application and where the
+  ``main()`` method runs on the cluster rather than the client. The job
+  submission is a one-step process: you don’t need to start a Flink cluster
+  first and then submit a job to the existing cluster session; instead, you
+  package your application logic and dependencies into a executable job JAR and
+  the cluster entrypoint ([ApplicationClusterEntryPoint]({{ site.baseurl
+  }}/api/java/index.html?org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.html))
+  is responsible for calling the ``main()`` method to extract the JobGraph.
+  This allows you to deploy a Flink Application like any other application on
+  Kubernetes, for example. The lifetime of a Flink Application Cluster is
+  therefore bound to the lifetime of the Flink Application.
+
+* **Resource Isolation**: in a Flink Application Cluster, the ResourceManager
+  and Dispatcher are scoped to a single Flink Application, which provides a
+  better separation of concerns than the Flink Session Cluster.
+
+<div class="alert alert-info"> <strong>Note:</strong> A Flink Job Cluster can be seen as a “run-on-client” alternative to Flink Application Clusters. </div>
+
+{% top %}
+
+## Self-contained Flink Applications
+
+When you want to do something like event-driven applications, it doesn’t make
+sense that you have to think about and manage clusters. So, there are efforts
+in the community towards enabling _Flink-as-a-Library_ in the future.
+
+The idea is that deploying a Flink Application becomes as easy as starting a
+process: Flink would be as any other library which you add to your application
+and does not affect how you deploy it. If you want to deploy such an
+application, it simply starts a set of processes which connect to each other,
+figure out their roles (e.g. JobManager, TaskManager) and execute the
+application in a distributed, parallel way. If the application cannot keep up
+with the workload, you simply start some new processes to rescale.
```

Comment:
Some small suggestions. I do question to what extent this should be positioned as something for the future, since this is largely possible today. If you want to keep it more firmly as something for the future, I suggest reworking the last sentence to talk more broadly about auto-scaling, rather than talking about manually scaling up.

```suggestion
When you want to create and deploy something like an event-driven application, it doesn’t make
sense that you have to think about and manage a cluster. So, there are efforts
in the community towards enabling _Flink-as-a-Library_ in the future.

The idea is that deploying a Flink Application becomes as easy as starting a
process: Flink would be like any other library which you add to your application,
with no effect on how you deploy it. When you want to deploy such an
application, it simply starts a set of processes which connect to each other,
figure out their roles (e.g. JobManager, TaskManager) and execute the
application in a distributed, parallel way. If the application cannot keep up
with the workload, you simply start some new processes to rescale.
```

Reply:
From what I understood in conversation with Stephan and Aljoscha, library mode is still "not there" yet. The idea here would mainly be to clear the myth that you always need a cluster to run Flink, and the progress of this effort can be discussed in the "Operations" docs. 

I do like your suggestion to hint at auto-scaling rather than implying that the processes need to be scaled manually and will rephrase some sentences!